File: code/4957/4957飞机降落.py
def dfs(u, last):
    if u == n:
        return True

    for i in range(n):
        a, b, c = t[i], d[i], l[i]
        if not st[i] and a + b >= last:
            st[i] = True
            if dfs(u + 1, max(last, a) + c):
                return True
            st[i] = False

    return False

if __name__ == "__main__":
    T = int(input())
    for _ in range(T):
        n = int(input())
        st = [False] * n
        t, d, l = [], [], []
        for _ in range(n):
            a, b, c = map(int, input().split())
            t.append(a)
            d.append(b)
            l.append(c)

        if dfs(0, 0):
            print("YES")
        else:
            print("NO")

File: code/4957/4957飞机降落.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
using namespace std;

const int N = 20;
bool st[N];
int T, n;

struct plane {
	int t, d, l;
} p[N];

bool dfs(int u, int last) {
	if (u == n) return true;

	for (int i = 0; i < n; i++) {
		int t = p[i].t, d = p[i].d, l = p[i].l;
		if (!st[i] && t + d >= last) {
			st[i] = true;
			if (dfs(u + 1, max(last, t) + l))
				return true;
			st[i] = false;
		}
	}

	return false;
}

int main() {
	cin >> T;
	while (T--) {
		cin >> n;
		for (int i = 0; i < n; i++) {
			int t, d, l;
			cin >> t >> d >> l;
			p[i] = {t, d, l};
		}

		memset(st, 0, sizeof st);
		if (dfs(0, 0)) puts("YES");
		else puts("NO");
	}

	return 0;
}
